### ISSUE TITLE
frozen layers if epoch < first_stage_epochs in train.py?

### ISSUE THREAD

**Issue by thomasmatt88** [hunglc007/tensorflow-yolov4-tflite#274]
I'm interested in transfer learning and therefore `first_stage_epochs` is set to 20. The model is frozen during training if `epoch < first_stage_epochs`. I assume this is so that the learning rate decreases to an appropriate level for transfer learning. It seems like a huge waste to train for 20 epochs while updating no weights. Am I missing something? If not, is there an `LR_INIT` and `LR_END` you would recommend specifically for transfer learning so that we can start updating weights right away?

**Comment by vladhondru25** [comment]
I think you update weights when `epoch < first_stage_epochs`, it's just the three output layers that are frozen. Correct me if I am wrong, but `freeze_layers` array only contains 3 layers.

**Comment by thegamercoder19** [comment]
In transfer learning, you take some part of high accuracy trained neural network and freeze it, e.g. in YOLOv4 we use yolov4.weights which was trained on COCO. This weights has trained long and its feature extractor doing well. So, it will be optimal to use well-doing feature extractor rather than train from scratch. But you don't need to freeze all model as class and bounding box head isn't right for your dataset. So, you train first 20 epochs class and bb head, after that you doing fine-tuning. File-tuning is advanced technique in transfer learning that allows you to get higher accuracy by unfreezing all model so feature extractor is fine-tuning by your dataset images. In fine-tuning, first epochs loss will be higher as you changing weights of well-doing feature extractor. But after some epochs it will descent as before fine-tuning feature extractor done well and so penalty for weights will become low.
